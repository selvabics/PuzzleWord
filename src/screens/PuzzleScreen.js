/**
 * This is puzzle game screen
 * Question will be load based on the level
 * Points will be provided based on the level
 * User needs to form the answer by selecting the given shuffled letters
 * User can see the points and level for the category they choose to play
 * User will be notified if the answere is wrong
 * If answer is right,they can share their points to their friends in social media
 * And they can move to next round if they wish to play next round
 */
import React, { useCallback, useEffect, useState } from 'react';
import { StyleSheet, FlatList, View, Text, TouchableOpacity, Alert } from 'react-native';
import { useDispatch, useSelector } from 'react-redux';
import * as Animated from 'react-native-animatable';

import CommonStyles from '../utils/styles';
import { Colors } from '../utils/constants';
import { pop } from '../router/navigation';
import { responsiveWidth } from 'react-native-responsive-dimensions';
import { getWord, getShuffledLetterList, getAnswerLetterList, playSound } from '../utils/utils';
import { getCurrentUserSelector } from '../redux/slice/loginSlice';
import { updateGameLevelAction } from '../redux/action/loginActions';
import { shareInSocialMediaAction } from '../redux/action/puzzleActions';
import BackgroundWallper from '../components/backgroundWallpaper';
import Header from '../components/header';
import ActionButton from '../components/actionButton';
import PuzzleBox from '../components/puzzleBox';
import Success from '../components/success';
import { SwapSound, WinSound, ErrorSound } from '../assets';

const PuzzleScreen = ({ route }) => {
    const dispatch = useDispatch();
    const { category } = route.params; // Selected category obj
    const puzzles = category.puzzles; // List of puzzle questions
    const currentUser = useSelector(getCurrentUserSelector); // Current logged in user data obj
    const [level, setLevel] = useState(currentUser.level[category.categoryId] || 0); // Game level of the selected category
    const [points, setPoints] = useState(currentUser.points[category.categoryId] || 0); // Scored points of the selected category
    const [question, setQuestion] = useState({}); // Store current question
    const [letters, setLetters] = useState([]); // Puzzle letters for reset purpose
    const [puzzleLetters, setPuzzleLetters] = useState([]); // Puzzle letters
    const [answerLetters, setAnswerLetters] = useState([]); // Answer letters
    const [error, setError] = useState(false); // Incorrect answer flag
    const [success, setSuccess] = useState(false); // Game level complete flag
    const puzzleColumn = Math.round(puzzles[level] ? puzzles[level]?.answer.length / 2 : 0);
    const [noOfColumn, setNoOfColumn] = useState(puzzleColumn > 5 ? 5 : puzzleColumn); // No of column the puzzle list should have 

    useEffect(() => {
        // Load puzzle questions
        loadQuestion();
    }, [level]);

    /**
     * This method is used to load question based on the level
     */
    const loadQuestion = () => {
        const question = puzzles[level];
        setQuestion(question);
        const shuffedLetters = getShuffledLetterList(question?.answer);
        setLetters(shuffedLetters);
        setPuzzleLetters(shuffedLetters);
        setAnswerLetters(getAnswerLetterList(question?.answer));
        const puzzleColumn = Math.round(question ? question?.answer?.length / 2 : 0);
        setNoOfColumn(puzzleColumn > 5 ? 5 : puzzleColumn);
    }

    /**
     * Navigate to previous screen
     */
    const goToHomeScreen = () => {
        pop();
    };

    /**
     * This method is used to go to next level of the game
     */
    const goToNextLevel = useCallback(() => {
        // If all the puzzles are solved in the category
        // Then show a success alert
        if (puzzles.length === level + 1) {
            Alert.alert('Success', 'You have completed all the levels in this category');
            return;
        }
        // Else move to next level
        setLevel(level + 1);
        setSuccess(false);
        loadQuestion();
    }, [level])

    /**
     * This method is called when user click on share button
     */
    const shareButtonClick = () => {
        // Share the points that you have scored to your friends in social media network
        const message = `I have scored ${question.points} in level ${level + 1} of ${category.categoryName}`;
        dispatch(shareInSocialMediaAction(message));
    }

    /**
     * This method is used to reset the letters to continue again
     */
    const resetLetters = () => {
        setPuzzleLetters([...letters]);
        setAnswerLetters(getAnswerLetterList(question?.answer));
    }

    /**
     * This method is used check whether the formed answer word is correct or not
     * If answer is correct then show success message on the screen
     * Else show error message
     * @param {String} answerWord 
     */
    const checkAnswerStatus = (answerWord) => {
        if (answerWord === question.answer) {
            setSuccess(true);
            playSound(WinSound);
            const payload = {
                categoryId: category.categoryId,
                level: level + 1,
                points: points + question.points,
            };
            dispatch(updateGameLevelAction(payload));
            setPoints(points + question.points);
        } else {
            setError(true);
            playSound(ErrorSound);
            setTimeout(() => {
                setError(false);
            }, 3000);
        }
    }

    /**
     * This method is used to move the selected letter to answer array
     * @param {Object} puzzleItem 
     * @param {Number} index 
     * @returns
     */
    const popLetterToAnswerList = (puzzleItem, index) => {
        // If no letter then do nothing
        if (!puzzleItem.letter) {
            return;
        }
        // Remove the selected letter from puzzle array
        const newPuzzleArr = [...puzzleLetters];
        newPuzzleArr[index] = {};
        setPuzzleLetters(newPuzzleArr);

        // Append the selected letter to answer array
        const newAnswerArr = [...answerLetters];
        const length = newAnswerArr.length;
        for (let i = 0; i < length; i++) {
            let item = newAnswerArr[i];
            if (!item?.letter) {
                newAnswerArr[i] = puzzleItem;
                setAnswerLetters(newAnswerArr);
                // Answer is formed in from user perspective
                // Need to check the answer word
                if (i === length - 1) {
                    // Get the answer word as string from array of letters
                    const answerWord = getWord(newAnswerArr);
                    checkAnswerStatus(answerWord);
                }
                break;
            }
        }
    };

    /**
     * This method is used to move back the or pop the last letter to answer array
     */
    const popLetterToPuzzleList = () => {
        playSound(SwapSound);
        let puzzleItem = {};
        const length = answerLetters.length;
        // Remove the letter from answer array and store it in a variable
        for (let i = length - 1; i >= 0; i--) {
            puzzleItem = answerLetters[i];
            if (puzzleItem?.letter) {
                const newAnswerArr = [...answerLetters];
                newAnswerArr[i] = {};
                setAnswerLetters(newAnswerArr);
                break;
            }
        }
        // Add the poped letter back to puzzle array
        if (puzzleItem) {
            const newPuzzleArr = [...puzzleLetters];
            newPuzzleArr[puzzleItem.index] = puzzleItem;
            setPuzzleLetters(newPuzzleArr);
        }
    };

    /**
     * This method renders success message with a button to share on social media
     * @returns 
     */
    const renderSuccess = () => {
        return (
            <View style={[CommonStyles.flex, CommonStyles.center]}>
                <Success points={question.points} />
                <ActionButton
                    style={styles.actionButton}
                    title={'SHARE'}
                    onPress={shareButtonClick}
                />
            </View>
        )
    }

    /**
     * This method renders the actual puzzle view with question and shuffled letters
     * @returns 
     */
    const renderPuzzle = () => {
        return (
            <View style={[CommonStyles.flex, CommonStyles.center]}>
                <TouchableOpacity onPress={popLetterToPuzzleList}>
                    <FlatList
                        key={`${noOfColumn}`}
                        scrollEnabled={false}
                        showsVerticalScrollIndicator={false}
                        contentContainerStyle={{ justifyContent: 'center', alignItems: 'center' }}
                        style={{ flexGrow: 0, flexWrap: 'wrap' }}
                        numColumns={noOfColumn}
                        data={answerLetters}
                        renderItem={({ item, index }) => (
                            <PuzzleBox item={item} />
                        )}
                    />
                </TouchableOpacity>
                <Animated.View animation={'bounceIn'}>
                    <TouchableOpacity style={{ margin: 30 }}>
                        <Text style={styles.questionText}>{question?.question}</Text>
                    </TouchableOpacity>
                </Animated.View>
                <FlatList
                    key={`${noOfColumn}`}
                    scrollEnabled={false}
                    showsVerticalScrollIndicator={false}
                    contentContainerStyle={{ justifyContent: 'center', alignItems: 'center' }}
                    style={{ flexGrow: 0, flexWrap: 'wrap' }}
                    numColumns={noOfColumn}
                    data={puzzleLetters}
                    renderItem={({ item, index }) => (
                        <PuzzleBox
                            item={item}
                            onPress={() => popLetterToAnswerList(item, index)}
                        />
                    )}
                />
            </View>
        );
    }

    return (
        <BackgroundWallper>
            <View style={CommonStyles.flex}>
                <Header
                    title={category.categoryName}
                    level={level + 1}
                    points={points}
                />
                <Text style={styles.errorText}>{error ? 'Incorrect Answer' : '.'}</Text>
                {success ? renderSuccess() : renderPuzzle()}
                <View style={CommonStyles.headerView}>
                    <ActionButton
                        style={styles.actionButton}
                        title={'BACK'}
                        onPress={goToHomeScreen}
                    />
                    {success ? (
                        <ActionButton
                            style={styles.actionButton}
                            title={'NEXT'}
                            onPress={goToNextLevel}
                        />) : (
                        <ActionButton
                            style={[styles.actionButton, { alignSelf: 'center' }]}
                            title={'RESET'}
                            onPress={resetLetters}
                        />
                    )}
                </View>
            </View>
        </BackgroundWallper>
    );
};

const styles = StyleSheet.create({
    errorText: {
        ...CommonStyles.h4,
        margin: 10,
        fontWeight: 'bold',
        textAlign: 'center',
        color: Colors.status,
    },
    questionText: {
        ...CommonStyles.h4,
        textAlign: 'center',
        color: 'white',
        fontWeight: 'bold',
        marginHorizontal: 20,
    },
    actionButton: {
        ...CommonStyles.rectBox,
        width: responsiveWidth(30),
    },
});

export default PuzzleScreen;
