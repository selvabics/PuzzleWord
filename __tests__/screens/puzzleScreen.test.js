import renderer from 'react-test-renderer';
import PuzzleScreen from '../../src/screens/PuzzleScreen';

import { Provider } from 'react-redux'
import configureStore from 'redux-mock-store';

jest.useFakeTimers();

describe('Testing login screen', () => {
    it('renders Login Screen', () => {
        const mockStore = configureStore()
        const currentUser = {
            email: '',
            level: {},
            points: {},
            totalPoints: 0,
        };
        const store = mockStore({ loginReducer: { currentUser } });
        const routeObj = { params: { category: { puzzles: []} } };

        // const store = mockStore({});
        const tree = renderer
            .create(<Provider store={store}><PuzzleScreen route={routeObj} /></Provider>)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
});

