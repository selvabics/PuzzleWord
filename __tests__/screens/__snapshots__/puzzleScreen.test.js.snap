// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing login screen renders Login Screen 1`] = `
<View
  accessibilityIgnoresInvertColors={true}
  style={
    [
      {
        "flex": 1,
        "paddingBottom": 20,
        "paddingTop": 30,
      },
      {},
    ]
  }
>
  <Image
    source={
      {
        "testUri": "../../../src/assets/images/background.jpg",
      }
    }
    style={
      [
        {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        {
          "height": undefined,
          "width": undefined,
        },
        undefined,
      ]
    }
  />
  <View
    style={
      {
        "flex": 1,
      }
    }
  >
    <View
      collapsable={false}
      forwardedRef={[Function]}
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "space-between",
          "opacity": 0,
          "padding": 20,
          "transform": [
            {
              "translateY": -800,
            },
          ],
          "width": "100%",
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            [
              {
                "fontSize": 22.94694968835727,
              },
              {
                "color": "white",
                "textAlign": "center",
              },
            ]
          }
        >
          Level
        </Text>
        <Text
          style={
            [
              {
                "fontSize": 45.89389937671454,
              },
              {
                "color": "white",
                "fontWeight": "bold",
                "textAlign": "center",
              },
            ]
          }
        >
          1
        </Text>
      </View>
      <Text
        style={
          {
            "color": "white",
            "fontSize": 45.89389937671454,
            "margin": 15,
            "textAlign": "center",
          }
        }
      />
      <View
        style={
          {
            "alignItems": "center",
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            [
              {
                "fontSize": 22.94694968835727,
              },
              {
                "color": "white",
                "textAlign": "center",
              },
            ]
          }
        >
          Scores
        </Text>
        <Text
          style={
            [
              {
                "fontSize": 45.89389937671454,
              },
              {
                "color": "white",
                "fontWeight": "bold",
                "textAlign": "center",
              },
            ]
          }
        >
          0
        </Text>
      </View>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <View
        collapsable={false}
        forwardedRef={[Function]}
        style={
          {
            "alignItems": "center",
            "alignSelf": "center",
            "backgroundColor": "rgba(136,213,66,0.8)",
            "borderRadius": 5,
            "borderWidth": 1,
            "height": 80.03999999999999,
            "justifyContent": "center",
            "margin": 20,
            "opacity": 0,
            "padding": 7.5,
            "transform": [
              {
                "scale": 0.3,
              },
            ],
            "width": 225,
          }
        }
      >
        <Text
          style={
            {
              "fontSize": 30.595932917809698,
              "fontWeight": "bold",
            }
          }
        >
          RESET
        </Text>
      </View>
    </View>
    <View
      style={
        [
          {
            "flex": 1,
          },
          {
            "alignItems": "center",
            "justifyContent": "center",
          },
        ]
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
          }
        }
      >
        <RCTScrollView
          contentContainerStyle={
            {
              "alignItems": "center",
              "justifyContent": "center",
            }
          }
          data={[]}
          getItem={[Function]}
          getItemCount={[Function]}
          keyExtractor={[Function]}
          onContentSizeChange={[Function]}
          onLayout={[Function]}
          onMomentumScrollBegin={[Function]}
          onMomentumScrollEnd={[Function]}
          onScroll={[Function]}
          onScrollBeginDrag={[Function]}
          onScrollEndDrag={[Function]}
          removeClippedSubviews={false}
          renderItem={[Function]}
          scrollEnabled={false}
          scrollEventThrottle={50}
          stickyHeaderIndices={[]}
          style={
            {
              "flexGrow": 0,
              "flexWrap": "wrap",
            }
          }
          viewabilityConfigCallbackPairs={[]}
        >
          <View />
        </RCTScrollView>
      </View>
      <View
        collapsable={false}
        forwardedRef={[Function]}
        style={
          {
            "opacity": 0,
            "transform": [
              {
                "scale": 0.3,
              },
            ],
          }
        }
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={false}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "margin": 30,
              "opacity": 1,
            }
          }
        >
          <Text
            style={
              {
                "color": "white",
                "fontSize": 38.244916147262124,
                "fontWeight": "bold",
                "marginHorizontal": 20,
                "textAlign": "center",
              }
            }
          />
        </View>
      </View>
      <RCTScrollView
        contentContainerStyle={
          {
            "alignItems": "center",
            "justifyContent": "center",
          }
        }
        data={[]}
        getItem={[Function]}
        getItemCount={[Function]}
        keyExtractor={[Function]}
        onContentSizeChange={[Function]}
        onLayout={[Function]}
        onMomentumScrollBegin={[Function]}
        onMomentumScrollEnd={[Function]}
        onScroll={[Function]}
        onScrollBeginDrag={[Function]}
        onScrollEndDrag={[Function]}
        removeClippedSubviews={false}
        renderItem={[Function]}
        scrollEnabled={false}
        scrollEventThrottle={50}
        stickyHeaderIndices={[]}
        style={
          {
            "flexGrow": 0,
            "flexWrap": "wrap",
          }
        }
        viewabilityConfigCallbackPairs={[]}
      >
        <View />
      </RCTScrollView>
    </View>
    <Text
      style={
        {
          "color": "rgb(111,240,205)",
          "fontSize": 38.244916147262124,
          "fontWeight": "bold",
          "margin": 20,
          "textAlign": "center",
        }
      }
    >
      .
    </Text>
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "space-between",
          "padding": 20,
          "width": "100%",
        }
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
          }
        }
      >
        <View
          collapsable={false}
          forwardedRef={[Function]}
          style={
            {
              "alignItems": "center",
              "backgroundColor": "rgba(136,213,66,0.8)",
              "borderRadius": 5,
              "borderWidth": 1,
              "height": 80.03999999999999,
              "justifyContent": "center",
              "margin": 20,
              "opacity": 0,
              "padding": 7.5,
              "transform": [
                {
                  "scale": 0.3,
                },
              ],
              "width": 225,
            }
          }
        >
          <Text
            style={
              {
                "fontSize": 30.595932917809698,
                "fontWeight": "bold",
              }
            }
          >
            BACK
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
